accounts-on-demand/src/subscription_manager.rs

```diff
@@ -0,0 +1,281 @@
+use std::{collections::HashMap, sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use itertools::Itertools;
+use solana_lite_rpc_accounts::account_store_interface::AccountStorageInterface;
+use solana_lite_rpc_cluster_endpoints::geyser_grpc_connector::GrpcSourceConfig;
+use solana_lite_rpc_core::{
+    commitment_utils::Commitment,
+    structures::{
+        account_data::{AccountData, AccountNotificationMessage, AccountStream},
+        account_filter::{AccountFilterType, AccountFilters, MemcmpFilterData},
+    },
+    AnyhowJoinHandle,
+};
+use solana_sdk::{account::Account, pubkey::Pubkey};
+use tokio::sync::{
+    broadcast::{self, Sender},
+    RwLock,
+};
+use yellowstone_grpc_proto::geyser::{
+    subscribe_request_filter_accounts_filter::Filter,
+    subscribe_request_filter_accounts_filter_memcmp::Data, subscribe_update::UpdateOneof,
+    SubscribeRequest, SubscribeRequestFilterAccounts, SubscribeRequestFilterAccountsFilter,
+    SubscribeRequestFilterAccountsFilterMemcmp,
+};
+
+pub struct SubscriptionManger {
+    accounts_filters: Arc<RwLock<AccountFilters>>,
+    restart_sender: broadcast::Sender<()>,
+}
+
+impl SubscriptionManger {
+    pub fn new(
+        grpc_sources: Vec<GrpcSourceConfig>,
+        accounts_storage: Arc<dyn AccountStorageInterface>,
+        account_notification_sender: Sender<AccountNotificationMessage>,
+    ) -> Self {
+        let (restart_sender, recver) = broadcast::channel(1);
+        let accounts_filters = Arc::new(RwLock::new(vec![]));
+
+        let (_, mut account_stream) =
+            create_grpc_account_streaming(grpc_sources, accounts_filters.clone(), recver);
+
+        let task_restart_sender = restart_sender.clone();
+        tokio::spawn(async move {
+            loop {
+                match tokio::time::timeout(Duration::from_secs(60), account_stream.recv()).await {
+                    Ok(Ok(message)) => {
+                        let _ = account_notification_sender.send(message.clone());
+                        accounts_storage
+                            .update_account(message.data, message.commitment)
+                            .await;
+                    }
+                    Ok(Err(e)) => match e {
+                        broadcast::error::RecvError::Closed => {
+                            log::error!("Account stream channel closed");
+                            break;
+                        }
+                        broadcast::error::RecvError::Lagged(_) => {
```

Comment:
uh, that should write a warning log, because account updates are effectively lost - no?